### ISSUE TITLE
Design - variable management

### ISSUE THREAD

**Issue by rachel-fenichel** [google/blockly#446]
Current plan: 
- [x] Explicit list of variables, updated on create/delete/rename, rather than implicit list created by walking the workspace.
- [x] "Make a variable" button in the flyout
- [x] "Make a list" button in the "lists" flyout
- [x] Variable type can be set when created
- [x] Variable type cannot be changed after creation
- [x] Variables category includes one getter for each existing variable and one of each setter block.  Currently there is one setter and one getter for each variable.

Necessary work:
- [x] Buttons in flyouts
- [x] Explicit variable list, with create/delete/rename interface
- [x] Variable creation dialog.  First pass: use existing dialog.  Second pass: update to allow setting type

Open questions:
- What types should be available to choose between by default?
- Are lists typed?
- How does a developer specify the available set of types
- How does a developer specify variables that they want to exist at start?
- Do we need the initial "item" variable now?

**Comment by pelikhan** [comment]
Great plan!

Sent from Outlook Mobile<https://aka.ms/djubve>

________________________________

**Comment by carlosperate** [comment]
> * Explicit list of variables, updated on create/delete/rename, rather than implicit list created by walking the workspace.
* "Make a variable" button in the flyout
* (likely) "Make a list" button in the "lists" flyout
* Variable type can be set when created
* Variable type cannot be changed after creation

To be completely honest I am not a big fan of the "make a typed variable" solution. I completely understand that is probably one of the best ways to ensure type definition, but one of the advantages of the original blockly implementation is not needing any guidance to be able to use variables. When users first start programming the concept of "making a variable" is not all that intuitive, it usually requires an explanation, and on top of that having type definitions will only confuse matters more and require further guidance.

On my own implementation the variable type is inherit from the first variable assignment, meaning that users could just drop a variable and start using it without knowing about types. It is not perfect, as you could end up in situations where the wrong block is attached (at the moment I display a block warning, the plan is to dynamically change each variable set input type to only accept the correct type) and it definitely has some drawbacks and quirks that I still need to work out, but I quite like the idea of being able to initially abstract this from the users, and perhaps offer explicit type definition as an slightly more "advanced" feature.

>Variable creation dialog. First pass: use existing dialog. Second pass: update to allow setting type

Hopefully here we can implement it in a synchronous manner that could be replaced using html modals and similar.


One other big question I would have with respect of this, how far the type definitions go. If we set a variable to represent a "number", are we stopping "non numeric" blocsk from being attached? Does that mean that all blocks will have a "type" defined in order to implement the input filter?

**Comment by rachel-fenichel** [comment]
More spec:
- Variables are no longer deleted when every use in the workspace disappears.
- Variables can be deleted by right clicking on the getter in the flyout and selecting "delete variable".  This deletes every getter and setter of that variable in the flyout
- Undo works on variable creation and deletion.

Necessary work:
- Allow right click on blocks in closeable flyouts
- Variable creation/deletion events.  I think also a variable rename event.

**Comment by rachel-fenichel** [comment]
> Variable creation dialog. First pass: use existing dialog. Second pass: update to allow setting type
> Hopefully here we can implement it in a synchronous manner that could be replaced using html modals and similar.

To clarify, first and second pass are referring to implementation passes.  That is, in the first implementation the "make a variable" button will just pop up the existing dialog, but you don't get there from the dropdown of an existing variable field.  Later it's easy to add typing, but there's no specific design yet.

>  Does that mean that all blocks will have a "type" defined in order to implement the input filter?

All connections already have "checks" defined.  Empty checks mean they can connect with anything.

With respect to typing, I would expect that for most simple applications, there would be no typing (hence the question of how the developer specifies which types show up as options; that could include no types).  And typed and non-typed variables could coexist in a single workspace.

A non-typed variable would probably act like our current ones do.  That is, they can accept anything, but no type information is preserved from one side to the other.  The only real alternative would be similar to the ternary block, which changes its output type based on its input type, possibly auto-disconnects, and is generally surprising in behaviour.

**Comment by carlosperate** [comment]
About the creation dialog I was just referring to https://github.com/google/blockly/pull/396#issuecomment-226893090.

> All connections already have "checks" defined. Empty checks mean they can connect with anything.

Yes, I guess we could thing of the output as the type.

> With respect to typing, I would expect that for most simple applications, there would be no typing (hence the question of how the developer specifies which types show up as options; that could include no types). And typed and non-typed variables could coexist in a single workspace.

But variables have no output defined. If you had something "set untyped variable x" =  "string block" you could then input the "x get block" into a "set variable y defined-as-a-number" block. If you can have "untyped variables" that do not inherit a type somehow, then they could go anywhere.

> A non-typed variable would probably act like our current ones do. That is, they can accept anything, but no type information is preserved from one side to the other. The only real alternative would be similar to the ternary block, which changes its output type based on its input type, possibly auto-disconnects, and is generally surprising in behaviour.

Unexpected auto-disconnects is the main reason why I've been a bit hesitant to implement more than block warnings at the moment for my fork, but they are not a good solution either as you can generate code that would not compile if you just ignore the warnings.
Most "errors" from blockly generated programs are basically user mistakes that might have originated from some ambiguity on how to use Blockly, but all Blockly generates code is valid code. This is perhaps something that needs to be taken in special consideration with this "type" implementation, as it could introduce the possibility of generating invalid code for new generators for statically typed languages or even some strongly typed languages.

**Comment by carlosperate** [comment]
> * Variables can be deleted by right clicking on the getter in the flyout and selecting "delete variable". This deletes every getter and setter of that variable in the flyout
>
> * Allow right click on blocks in closeable flyouts

Is this intuitive enough? There are already a few option only available through the right click menu, but I haven't seen many people even trying to use the right click and I've talked to some users with a relative good experience using blockly and that were not aware of some of these options at all.

This would be the first instance of this particular pattern, no? I cannot speak for others, but I know that I probably wouldn't look to a right context menu on the flyout until I've exhausted all other plausible options. But that might just be me, I usually struggle with finding these type of options on software.

**Comment by rachel-fenichel** [comment]
> Is this intuitive enough? There are already a few option only available through the right click menu, but I haven't seen many people even trying to use the right click and I've talked to some users with a relative good experience using blockly and that were not aware of some of these options at all.

Yes, we discussed this today and concluded that it's reasonable--not because it's intuitive enough to be found, but because deleting an already-unused variable is not a core user action.

Right click on a non-closeable flyout shows "help" now.  On a closeable flyout it does nothing.  This would make it show "help" for just about everything, and then "delete a variable" as an option on the variable getters.

**Comment by rachel-fenichel** [comment]
A non-typed variable would accept anything in its setters and connect to anything with its getter.  A typed variable would have types on both its getters and its setters.  I don't understand what you mean by 
>But variables have no output defined. 

But dynamically changing the type of a variable based on things connected to it, especially if it's already in use throughout the workspace, necessarily leads either to auto-disconnects or a need for warnings.  Choosing the type (or no type/all types) at the beginning solves this.

**Comment by carlosperate** [comment]
By "no output defined" I mean that a normal variable does not have an output type defined in the block: https://github.com/google/blockly/blob/master/blocks/variables.js#L48

>But dynamically changing the type of a variable based on things connected to it, especially if it's already in use throughout the workspace, necessarily leads either to auto-disconnects or a need for warnings. Choosing the type (or no type/all types) at the beginning solves this.

Yes, I am not disputing this at all, but I am saying that unless variables start updating their output type, we could have situations like the following (let say x is a "string-type variable", y is a "non-typed", and "[ ]" represents a block):

```
[set x to ["hello world!"] ]
[set y to [x] ]
```

**Comment by rachel-fenichel** [comment]
this.setOutput(true) is the same as this.setOutput(true, null).  So there's implicitly an output type, it's just the wildcard.

Re: your example: interesting.  Basically you're saying you could coerce any variable into any other by passing it through an intermediate non-typed variable, right?

So say x is a "number-type variable".  Then I could do
```
[set z to [10]
[set y to [z]]
[set x to [y]]
```

and it would be valid.  Is that your concern?

**Comment by carlosperate** [comment]
> Re: your example: interesting. Basically you're saying you could coerce any variable into any other by passing it through an intermediate non-typed variable, right?

Yes, although in your example:
```
[ set z to [10] ]
[ set y to [z] ]
[ set x to [y] ]
```

It would still be correct, if x is a "number-type variable", since z is set to a number. But let's say z is defined as a "number-type variable" and x would instead be defined to a "string-type variable", in that case because of `y` you would have a "string-type x = number-type z" .

**Comment by rachel-fenichel** [comment]
I meant z is number-type.  Sorry.  Will edit the comment.

**Comment by NeilFraser** [comment]
> What types should be available to choose between by default?
> How does a developer specify the available set of types

I'd imagined that this should just be a [configuration option](https://developers.google.com/blockly/guides/get-started/web#configuration):

types:	array	List of strings, each defining an allowed variable type (or function return value). Defaults to empty list which enables untyped variables.

**Comment by NeilFraser** [comment]
> Explicit list of variables, updated on create/delete/rename, rather than implicit list created by walking the workspace.

I think walking the workspace will still be needed:
* Create 'foo' variable.
* With a 'foo' block: create, copy, delete.
* Delete 'foo' variable.
* Paste.

**Comment by rachel-fenichel** [comment]
@NeilFraser is that a sequence of events that you think could be problematic?

Should pasting a block with a non-existent variable selected in a dropdown create the variable?

**Comment by rachel-fenichel** [comment]
See the [feature/variable_management](https://github.com/google/blockly/tree/feature/variable_management) branch for progress on this so far.

**Comment by rachel-fenichel** [comment]
If you're interesting in this topic, please take a look at feature/variable_management and file bugs against me.

In particular,
 - Copy/paste
 - Variable deletion
 - Undo/redo

This branch does not include variable typing, but is a prerequisite.  My goal is to get this branch fully stable and then merged into master before adding typing.

**Comment by RoboErikG** [comment]
Additional feature for discussion:
When deleting a variable if it's not the last instance ask the user for confirmation and specify how many blocks using the variable will be deleted. This gives users an opportunity to clean up usages of the variable before deleting it.

**Comment by geppy** [comment]
@rachel-fenichel This looks great.  I'd be interested in helping out.  Would you mind checking off the bits that've been completed in [the TODO list](https://github.com/google/blockly/issues/446#issue-161508455)?  I'd like to send you a PR to fix #493, but I'm not sure how function parameters fit into your design.

**Comment by rachel-fenichel** [comment]
@geppy TODO list updated!  Note that it refers to what's in the variable management branch, not what's in master.
Function parameters: I have to think about the design, but will try to have an answer today or tomorrow.

Thanks!
